saleor/product/migrations/0183_calculate_discounted_prices.py

```diff
@@ -0,0 +1,395 @@
+    collections = (
+        sale_model.collections.through.objects.filter(sale_id__in=sale_pks)
+        .order_by("id")
+        .values_list("sale_id", "collection_id")
+    )
+    collection_map = defaultdict(set)
+    for sale_pk, collection_pk in collections:
+        collection_map[sale_pk].add(collection_pk)
+    return collection_map
+
+
+def fetch_products(sale_model, sale_pks):
+    products = (
+        sale_model.products.through.objects.filter(sale_id__in=sale_pks)
+        .order_by("id")
+        .values_list("sale_id", "product_id")
+    )
+    product_map = defaultdict(set)
+    for sale_pk, product_pk in products:
+        product_map[sale_pk].add(product_pk)
+    return product_map
+
+
+def fetch_variants(
+    sale_model,
+    sale_pks,
+):
+    variants = (
+        sale_model.variants.through.objects.filter(sale_id__in=sale_pks)
+        .order_by("id")
+        .values_list("sale_id", "productvariant_id")
+    )
+    variants_map = defaultdict(set)
+    for sale_pk, variant_pk in variants:
+        variants_map[sale_pk].add(variant_pk)
+    return variants_map
+
+
+def fetch_sale_channel_listings(
+    sale_listing_model,
+    sale_pks,
+):
+    channel_listings = sale_listing_model.objects.filter(sale_id__in=sale_pks)
+    channel_listings_map = defaultdict(dict)
+    for channel_listing in channel_listings:
+        sale_id_row = channel_listings_map[channel_listing.sale_id]
+        sale_id_row[channel_listing.channel_id] = channel_listing
+    return channel_listings_map
+
+
+def fetch_categories(
+    sale_model,
+    category_model,
+    sale_pks,
+):
+    categories = (
+        sale_model.categories.through.objects.filter(sale_id__in=sale_pks)
+        .order_by("id")
+        .values_list("sale_id", "category_id")
+    )
+    category_map = defaultdict(set)
+    for sale_pk, category_pk in categories:
+        category_map[sale_pk].add(category_pk)
+    subcategory_map = defaultdict(set)
+    for sale_pk, category_pks in category_map.items():
+        subcategory_map[sale_pk] = set(
+            category_model.tree.filter(pk__in=category_pks)
+            .get_descendants(include_self=True)
+            .values_list("pk", flat=True)
+        )
+    return subcategory_map
+
+
+def update_discounted_prices_task(
+    product_model,
+    collection_product_model,
+    product_listing_model,
+    variant_listing_model,
+    variant_model,
+    discounts,
+    start_pk: int = 0,
+):
+    products = list(
+        product_model.objects.filter(pk__gt=start_pk)
+        .prefetch_related("channel_listings", "collections")
+        .order_by("pk")[:BATCH_SIZE]
+    )
+
+    if products:
+        update_products_discounted_price(
+            product_model,
+            collection_product_model,
+            product_listing_model,
+            variant_listing_model,
+            variant_model,
+            products,
+            discounts,
+        )
+        update_discounted_prices_task(
```

Comment:
I would suggest to test this migration on some bigger dataset to check if we can manage that recursion.

Reply:
I tested it on a DB with 1.5 M of products and it goes smoothly, each `update_discounted_prices_task` is run for about 1 second.